packages/contracts/src/PotLuck.sol

```diff
@@ -0,0 +1,199 @@
+            p.totalParticipants++; 
+        }
+
+
+        IERC20(p.token).safeTransferFrom(msg.sender, address(this), p.entryAmount);
+        p.balance += p.entryAmount;
+        p.participants.push(msg.sender);
+        
+        hasJoinedRound[joinKey] = true;
+
+        emit Joined(potId,p.round, msg.sender);
+    }
+
+    /**
+     * @notice Trigger payout for pot after deadline; picks a random winner, sends pot minus fee.
+     *         Resets round for next period.
+     * @param potId Identifier of the pot.
+     */
+    function triggerPayout(uint256 potId) external {
+        Pot storage p = pots[potId];
+    require(p.balance > 0,                "Pot not exist");
+    require(block.timestamp >= p.deadline, "Round in progress");
+
+    // Detect “last round”: we’ve already done totalParticipants−1 rounds (0-indexed)
+   bool isLast = (p.round == p.totalParticipants - 1);
+
+   // If NOT last round, we keep one entryAmount as rollover
+    uint256 rollover = isLast ? 0 : p.entryAmount;
+    uint256 total    = p.balance;
+    uint256 fee      = (total * platformFeePct) / 100;
+    require(total >= fee + rollover, "Insufficient funds to rollover");
+
+    uint256 prize = total - fee - rollover;
+
+    bytes32 seed = keccak256(
+        abi.encodePacked(
+            blockhash(block.number - 1),
+            potId,
+            p.round,
+            total
+        )
+    );
+
+    // 2) pick a start index in [0..len)
+    uint256 len = p.participants.length;
+    uint256 idx = uint256(seed) % len;
+
+    // 3) scan forward until we find someone !hasWon
+    address winner;
+    for (uint256 i = 0; i < len; i++) {
+        address candidate = p.participants[(idx + i) % len];
+        bytes32 candidateKey = keccak256(abi.encodePacked(potId, candidate));
+
+        if (!hasWon[candidateKey]) {
+            winner = candidate;
+            break;
+        }
+    }
+
+    require(winner != address(0), "No eligible participants");
+
+    bytes32 winnerKey = keccak256(abi.encodePacked(potId, winner));
+    hasWon[winnerKey] = true;
+
+
+    IERC20(p.token).safeTransfer(winner, prize);
+    if (fee > 0) {
+        IERC20(p.token).safeTransfer(owner(), fee);
+    }
+
+    uint32 nextRound = ++p.round;
+
+    emit Payout(potId, winner, prize, nextRound - 1);
+    if (isLast) {
+       // --- FINAL ROUND: no more rollover, end the pot ---
+       p.balance = 0;
+       delete p.participants;
+       p.isPublic = false;
+       return;
+   }
+
+    p.balance = rollover;
+    delete p.participants;
+
+    // auto-re-enter winner into new round
+    p.participants.push(winner);
+    bytes32 nextKey = keccak256(abi.encodePacked(potId, nextRound, winner));
+    hasJoinedRound[nextKey] = true;
+    emit Joined(potId, nextRound, winner);
+
+    p.deadline = block.timestamp + p.period;
+    }
+
+    /** @notice Allows owner to update platform fee percentage (<=100). */
+    function setPlatformFeePct(uint256 pct) external onlyOwner {
```

Comment:
I think a flat platform fee is better to start. Eg always 1 USDC. Don't want to discourage people from making big pots at this stage.